src/main/java/it/unisa/Amigo/consegna/services/ConsegnaServiceImpl.java

```diff
@@ -1,4 +1,196 @@
+import org.springframework.web.multipart.MultipartFile;
+
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Questa classe implementa i metodi  per la logica di Business del sottositema "Consegna"
+ */
+@Service
+@RequiredArgsConstructor
+public class ConsegnaServiceImpl implements ConsegnaService {
+
+    private final ConsegnaDAO consegnaDAO;
+
+    private final DocumentoService documentoService;
+
+    private final GruppoService gruppoService;
+
+    /**
+     * Effettua la consegna di un documento ad uno o più destinatari
+     *
+     * @param idDestinatari gli id dei destinatari che riceveranno il documento
+     * @param file          il file allegato al documento
+     */
+    @Override
+    public List<Consegna> sendDocumento(int[] idDestinatari, String locazione, MultipartFile file) {
+        Documento doc = documentoService.addDocumento(file);
+        List<Consegna> result = new ArrayList<>();
+        if (idDestinatari != null) {
+            for (int id : idDestinatari) {
+                Consegna consegna = new Consegna();
+                consegna.setDataConsegna(LocalDate.now());
+                consegna.setStato("NO_VALUTARE");
+                consegna.setDocumento(doc);
+                consegna.setMittente(gruppoService.getAuthenticatedUser());
+                consegna.setLocazione(Consegna.USER_LOCAZIONE);
+                consegna.setDestinatario(gruppoService.findPersona(id));
+                result.add(consegna);
+                consegnaDAO.save(consegna);
+            }
+        } else {
+            Consegna consegna = new Consegna();
+            consegna.setDataConsegna(LocalDate.now());
+            consegna.setStato("DA_VALUTARE");
+            consegna.setDocumento(doc);
+            consegna.setMittente(gruppoService.getAuthenticatedUser());
+            if (locazione.equalsIgnoreCase(Consegna.PQA_LOCAZIONE))
+                consegna.setLocazione(Consegna.PQA_LOCAZIONE);
+            if (locazione.equalsIgnoreCase(Consegna.NDV_LOCAZIONE))
+                consegna.setLocazione(Consegna.NDV_LOCAZIONE);
+            result.add(consegna);
+            consegnaDAO.save(consegna);
+        }
+        return result;
+    }
+
+    /**
+     * Effettua il download di un documento
+     *
+     * @param idDocument il documento da scaricare
+     * @return la pagina di visualizzazione del documento scaricato
+     */
+    @Override
+    public ResponseEntity<Resource> downloadDocumento(int idDocument) {
+        Documento documento = documentoService.findDocumentoById(idDocument);
+        Resource resource = documentoService.loadAsResource(documento);
+
+        return ResponseEntity.ok()
+                .contentType(MediaType.parseMediaType(documento.getFormat()))
+                .header(HttpHeaders.CONTENT_DISPOSITION, "filename=\"" + documento.getNome() + "\"")
+                .body(resource);
+    }
```

Comment:
Questa non è competenza di un service. Response entity è roba che riguarda il web layer e quindi va in un controller. Suggerisco di togliere il metodo dal service. Basta unicamente un metodo che dal service permetta di recuperare il documento tramite id (che credo già ci sia) e poi il metodo che crea la response entity va messo nel Consegna Controller.
**Se il metodo di download (che restituisce una response entity) è stato messo anche su altri service va corretto anche negli altri service. Creo una issue su github**.